# Mushroom Phylogeny

This lab on Evolutionary Patterns course has been inspired by [this](https://github.com/iirisarri/PEB_Phylogenomics) guide on phylogenomics.

This document will guide you through phylogenetics pipeline—–allowing you to go from a bunch of sequences to a phylogenetic tree that represents how the sequences are related to each other. The phylogenetic analyses will yield a phylogenetic hypothesis in terms of shared history (tree topology), but also in terms of evolutionary change (branch lengths) and other
interesting parameters such as relative substitution rates between loci, base frequencies, relative substitution frequencies (between different bases), etc.

The phylogenomics pipeline can become very complex, many additional steps might be included (particularly at the satage of dataset assembly!) and some analyses can take weeks to complete. Pipelines are modular, meaning they can (and should) be improved, as well as modified to the particular question at hand.

## Objective and data

The objective of this lab is to use the sequence reads that you generated in conjunction with existing data in GenBank/SILVA for phylogenetic analyses of the mushroom group that you were assigned. This practical can be roken down into several steps:
1) Trimming Sanger sequences and assembling contigs
2) Assembling data sets for each locus based on your own sequences (above) and entries in GenBank/SILVA
3) Phylogenetic analyses of separate and combined data sets, using (a) Parsimony and (b) Bayesian methods 
4) Evaluating the results 

## Programs required
1) A simple text editor to view and edit files in plain text format. I recommend [TextWrangler or BBEdit](https://www.barebones.com/products/textwrangler/) for Mac and [Notepad++](https://notepad-plus-plus.org/downloads/) for Windows.
2) [Assseq](https://ormbunkar.se/assseq/downloads/), a programme for processing Sanger sequences  

## 1. Trimming Sanger sequences and assembling contigs

The first step is to process the Sanger sequences that you generated. This involves: 
- calling bases/processing a chromatogram 
- trimming poor quality sequence
- merging the forward and reverse read together into a single consensus sequence or contig

We will use [Assseq](https://github.com/andersla/Assseq) for this task which is simple to use. 

Start by downloading your sequences from StudentPortalen (**Documents -> Labs -> Mushroom Phylogeny -> Sanger sequences**). Only download the sequences generated by your group (check these in the Mushrooms&Labelling.xlsx file on StudentPortalen). Next, load the forward and reverse reads in .ab1 format, one sequence at a time in Assseq. Scroll to view the chromatograms. *Do you have good quality sequences?* 

Trim poor quality bases and assemble into a contig (**Assemble -> Reassemble everything with CAP3 external assembler plugin**). You can now edit the consensus sequence (only if necessary) or reverse complement it if it is in the incorrect orientation. Save the consensus sequence (**File -> Save consensus as Fasta**) (be sure to give it a good name - the name of the mushroom and the gene).

*Imp*: Verify that you actually did amplify a mushroom and the correct gene by [BLASTing](https://blast.ncbi.nlm.nih.gov/Blast.cgi) your consensus sequence.


## 2. Assembling datasets for each locus

To build your phylogenies, you will need sequences from other taxa in your assigned group. This is a non-trivial task and it can be time consuming to assemble your dataset. 

There are several good starting points for building your datasets:
1) The NCBI nucleotide database (GenBank)——an open access, annotated collection of all publicly available nucleotide sequences. Use this to find sequences for 28S/LSU, RPB1 and RPB2. The [Taxonomy Browser](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi) is a good place to start. Simply search for your group of interest and click on the the *Nucleotide* records. 

2) The [SILVA database](https://www.arb-silva.de/) which is a database of ribosomal genes (18S and 28S). 

3) Published papers! Search on [Google Scholar](https://scholar.google.com/) for phylogenetic studies of your mushroom group. This can point you to accession numbers of sequences used in previous studies, and is good for ensuring that you have a good taxon sampling. 

If possible, try to download sequences for the three loci from the same species. It is probable that sequences for one locus will be easier to find. 
**NB: Don't forget to retrieve sequences for the outgroup!**


The output from both the contig assembly (your own sequences) and the sequences you have downloaded from online databases should be stored as individual FASTA files (see below for description of the fasta format). You will have to edit the header to include only the information you are interested in. You will need both the species names, and the accession number for later use (in your report). Make sure that you have not edited any sequences using TextEdit/Microsoft Word etc. wich saves the files as RTF/doc/something else rather than as text files (use TextWrangler/BBEdit/Notepad++ for editing). 

<details>
<summary>What is a fasta file?</summary>
<br>
FASTA format is a text-based format for representing either nucleotide sequences or peptide sequences. A sequence in FASTA format consists of:
- Header: One line starting with a ">" sign, followed by a sequence identification code.
- Sequence line(s): One or more lines containing the sequence itself.
  
Example of 2 sequences in FASTA format:

>sequence A
ggtaagtcctctagtacaaacacccccaatattgtgatataattaaaattatattcatat
tctgttgccagaaaaaacacttttaggctatattagagccatcttctttgaagcgttgtc
>sequence B
ggtaagtgctctagtacaaacacccccaatattgtgatataattaaaattatattcatat
tctgttgccagattttacacttttaggctatattagagccatcttctttgaagcgttgtc
tatgcatcgatcgacgactg
</details>



